src/ripple/overlay/impl/MessageScheduler.cpp

```diff
@@ -0,0 +1,413 @@
+#include <ripple/basics/Log.h>
+#include <ripple/basics/chrono.h>
+#include <ripple/overlay/MessageScheduler.h>
+
+#include <cassert>
+#include <chrono>
+#include <exception>
+#include <iterator>
+#include <mutex>
+
+namespace ripple {
+
+using RequestId = MessageScheduler::RequestId;
+
+// We use a thread-local variable to track whether a sender is being
+// scheduled from a callback in a thread where the scheduler is already
+// locked. In that case, trying to lock the scheduler again is undefined
+// behavior. Instead, we just save the sender here to be handled after the
+// callback returns.
+thread_local static std::vector<MessageScheduler::Sender*>* tsenders = nullptr;
+// `caller` is logged whenever a sender is scheduled or a channel is offered.
+// It names the method that called either (a) the callback that called
+// `schedule` or (b) `_offer`.
+// The (b) case can be solved with a parameter (to `offer_`), but the (a) case
+// cannot, and if we're using a thread-local variable for the (a) case anyway,
+// it doesn't hurt anything to use it for the (b) case too.
+thread_local static char const* caller = "none";
+
+/**
+ * "Push" a value at this point in the call stack.
+ * "Pop" the value when exiting the current scope.
+ */
+template <typename T>
+struct push_value
+{
+    T& variable;
+    T previous;
+
+    push_value(T& variable, T next)
+        : variable(variable), previous(std::move(variable))
+    {
+        variable = std::move(next);
+    }
+
+    ~push_value()
+    {
+        variable = std::move(previous);
+    }
+};
+
+void
+MessageScheduler::connect(std::vector<Channel> channels)
```

Comment:
Nit. This could be
`MessageScheduler::connect(std::vector<Channel>& channels)` or `MessageScheduler::connect(std::vector<Channel>&& channels)` to avoid a copy.

Reply:
This was intentional, but I can change it to rvalue-reference.

Reply:
See [additional comment](https://github.com/XRPLF/rippled/pull/4217#discussion_r1151015762).